# Project 1: Profitable App Profiles for the App Store and Google Play Markets


## What is the goal?
To understand what type of apps are likely to attract more users on Google Play and App Store.

There are two sets of data (downloaded from Kaggle). 
1. [A data set](https://www.kaggle.com/lava18/google-play-store-apps) containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018. 

2. [A data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017. 

Because our end goal is to develop and add an app that's going to be profitable on both Google Play and the App Store, we need to find app profiles that are successful on both markets. (For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.)

In this project, I'm taking different ways to clean and eye-ball the data, and examine the popularity of particular kinds of apps using frequency tables. 

# 1. Explore data
Below, I read in the data using two different ways: First, I use `pandas` then I read in the data without it. In the future, I can use either option.

## Option 1: Read in the data using `pandas`


In [10]:
import pandas as pd

# apple app data
apple_df = pd.read_csv("AppleStore.csv")
apple_df.head()

id               track_name  size_bytes currency  price  \
0  284882215                 Facebook   389879808      USD    0.0   
1  389801252                Instagram   113954816      USD    0.0   
2  529479190           Clash of Clans   116476928      USD    0.0   
3  420009108               Temple Run    65921024      USD    0.0   
4  284035177  Pandora - Music & Radio   130242560      USD    0.0   

   rating_count_tot  rating_count_ver  user_rating  user_rating_ver      ver  \
0           2974676               212          3.5              3.5     95.0   
1           2161558              1289          4.5              4.0    10.23   
2           2130805               579          4.5              4.5  9.24.12   
3           1724546              3842          4.5              4.0    1.6.2   
4           1126879              3594          4.0              4.5    8.4.1   

  cont_rating        prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
0          4+  Social Networking               37                1        29   
1         12+      Photo & Video               37                0        29   
2          9+              Games               38                5        18   
3          9+              Games               40                5         1   
4         12+              Music               37                4         1   

   vpp_lic  
0        1  
1        1  
2        1  
3        1  
4        1

In [11]:
# google app data
google_df = pd.read_csv("googleplaystore.csv")
google_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

##  Option 2: Read in the data without `pandas` (using a different function)

Below is a function named explore_data() that I can repeatedly use to print rows in a readable way.


In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

The `explore_data()` function:

I. Takes in four parameters:
1. `dataset`, which is expected to be a list of lists.
2. `start` and `end`, which are both expected to be integers and represent the starting and the ending indices of a slice from the data set.
3. `rows_and_columns`, which is expected to be a Boolean and has `False` as a default argument.

II. Slices the data set using `dataset[start:end]`.

III. Loops through the slice, and for each iteration, prints a row and adds a new line after that row using print('\n').

IV. Prints the number of rows and columns if rows_and_columns is True.
* `dataset` shouldn't have a header row, otherwise the function will print the wrong number of rows (one more row compared to the actual length).

In [3]:
## Open the AppleStore data set 
open_apple = open('AppleStore.csv')
from csv import reader
read_apple = reader(open_apple)
apple_data = list(read_apple)

# create the list of the header row 
apple_data_header = apple_data[0]

# the rest of the data set
apple_data = apple_data[1:]

# Take a look at the header 
# Use the explore_data() funciton to explore rows in a readable way.
print(apple_data_header)
print('\n')
explore_data(apple_data, 0, 3, True) # print the header row + first 3 rows.

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


The Apple Store data set has 7197 entries (i.e., apps) and 16 columns. 

In [12]:
## Open the Google Play Store data set 
open_google = open('googleplaystore.csv')
read_google = reader(open_google)
google_data = list(read_google)


# create the list of the header row 
google_data_header = google_data[0]

# the rest of the data set
google_data = google_data[1:]

# Take a look at the header 
# Use the explore_data() funciton to explore rows in a readable way.
print(google_data_header)
print('\n')
explore_data(google_data, 0, 3, True) # print the header row + first 3 rows.

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


The Google Play Store data set has 10841 entries (i.e., apps) and 13 columns.

# 2. Clean data - deal with rows with missing information.
The Google Play data set has a dedicated discussion section, and [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472. Let's print this row and compare it against the header and another row that is correct.

In [4]:
# Print the header row, the row at that index number, and the correct row (first row) to check whether it's indeed incorrect. 
print(google_data_header)
print(google_data[10472])
print(google_data[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


In [13]:
# I could find out which row has potentially a different number of columns than the header row.
for row in google_data:
    if len(row) != len(google_data_header):
        print(row)
        print(google_data.index(row))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10472


Based on the info above, the row for the app called 'Life Made WI-Fi Touchscreen Photo Frame' has different number of columns than other rows; 'Category' is missing, and the 'Rating' value is 19 (the maximum should be 5). So we'll remove this row. 

In [14]:
del google_data[10472] # don't run this code more than once. 

In [15]:
len(google_data) # now the google data set has 10840 apps. 

10840

# 3. Clean data - deal with duplicate rows.

One of the discussions suggests that there are duplicate entries. 

First, find out how many unique apps are in the data set, and how many rows are duplicate. 

In [16]:
unique_apps = []
duplicate_apps = []

for row in google_data:
    name = row[0] # specify that the first column in each row is the apps name.
    
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print("number of unique apps:", len(unique_apps)) # this shows that there are 9659 unique apps.
print('\n')
print("number of duplicate rows:", len(duplicate_apps)) # this shows that there are 1181 instances of duplicate entries. 

number of unique apps: 9659


number of duplicate rows: 1181


In [17]:
print("examples of duplicate rows:", duplicate_apps[0:10]) # take a look at the first few entries of duplicate instances.

examples of duplicate rows: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


In [18]:
# For example, we can see that there are three rows of 'Slack'.
for row in google_data:
    name = row[0]
    
    if name == 'Slack':
        print(row)

['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


I want the data set to have only one entry per app. But how do I select which entry to leave in the data vs. delete? 

If I examine the duplicate rows for 'Slack", I can see that the fourth column has different numbers (i.e., for the number of Reviews). Rather than deleting duplicate rows randomly, I'm going to pick the row that has the highiest number of reviews because the more reviews there are the more reliable the rating is. 

To remove the duplicates, I will:

1. Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.
2. Use the information stored in the dictionary and create a new data set, which will have only one entry per app (and for each app, I'll only select the entry with the highest number of reviews).

Step 1: Create a dictionary where each key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.

In [19]:
reviews_max = {}

for row in google_data:
    name = row[0]
    n_reviews = float(row[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max: 
        reviews_max[name] = n_reviews

len(reviews_max) # the expected number here is 9659 (i.e., the number of unique apps)

9659

Step 2: Use the dictionary I created above to remove the duplicate rows.


1. Start by initializing two empty lists, android_clean and already_added.
2. Loop through the android (google) data set, and for every iteration:
    - isolate the name of the app and the number of reviews.
    - add the current row (app) to the android_clean list, and the app name (name) to the already_added list if:
        - The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and 
        - The name of the app is not already in the already_added list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [20]:
android_clean = [] # here store the new cleaned data set
already_added = [] # here store just app names.

for row in google_data:
    name = row[0]
    n_reviews = float(row[3])
    
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        android_clean.append(row)
        already_added.append(name)
        
# Step 3
len(android_clean) # there should be 9659 unique entries. 

9659

Just in case, make sure that there are no duplicate rows for the Apple store data. 

In [21]:
unique_apps_apple = []
duplicate_apps_apple = []

for row in apple_data:
    name = row[0] # specify that the first column in each row is the apps name.
    
    if name in unique_apps_apple:
        duplicate_apps_apple.append(name)
    else:
        unique_apps_apple.append(name)
        
print("number of Apple unique apps:", len(unique_apps_apple)) # this shows that there are 7197 unique apps.
print('\n')
print("number of Apple duplicate rows:", len(duplicate_apps_apple)) # this shows that there 0 instances of duplicate entries. 

number of Apple unique apps: 7197


number of Apple duplicate rows: 0


## 4. Clean data - deal with apps in non-English language.
In this particular project, I want to analyze only the apps that are directed toward an English-speaking audience. However, there are some apps that do not have English names.

In [22]:
print(apple_data[813][1])
print(apple_data[6731][1])

print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system. Based on this number range, I can build a function that detects whether a character belongs to the set of common English characters or not.

In [23]:
print(ord('a'))
print(ord('+'))
print(ord('6'))
print(ord('あ'))
print(ord('書'))

97
43
54
12354
26360


This is a a function that takes in a string and returns `False` if there's any character in the string that doesn't belong to the set of common English characters, otherwise it returns `True`.

In [24]:
def English_char(string):
    for character in string:
        if ord(character) > 127:
            return False
    
    return True

# This function can be used like this:
print(English_char('Instagram'))
print(English_char('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(English_char('Docs To Go™ Free Office Suite'))
print(English_char('Instachat 😜'))

True
False
False
False


Emojis and characters like ™ fall outside the ASCII range and have corresponding numbers over 127.

In [25]:
print(ord('™'))
print(ord('😜'))

8482
128540


To minimize the impact of data loss, I'll only remove an app if its name has more than three non-ASCII characters. If the input string has more than three characters that fall outside the ASCII range (0 - 127), then the function should return `False` (identify the string as non-English), otherwise it should return `True`

In [26]:
def English_char(string):
    non_eng_char = 0
    
    for character in string:
        if ord(character) > 127:
            non_eng_char += 1
            
    if non_eng_char > 3:
        return False
    else:
        return True

print(English_char('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(English_char('Docs To Go™ Free Office Suite'))
print(English_char('Instachat 😜'))

False
True
True


I now use the new function to filter out non-English apps from both data sets. Loop through each data set. If an app name is identified as English, append the whole row to a separate list. 

In [27]:
# Apple data
apple_data_eng = []
apple_data_non_eng = []

for row in apple_data:
    name = row[1]
    
    if English_char(name) == False:
        apple_data_non_eng.append(row)
    else:
        apple_data_eng.append(row)

print("the number of Apple apps in English:", len(apple_data_eng))
print("the number of Apple apps in non-English:", len(apple_data_non_eng))

the number of Apple apps in English: 6183
the number of Apple apps in non-English: 1014


In [28]:
# Google data 
google_data_eng = []
google_data_non_eng = []

for row in android_clean:
    name = row[0] # name column is different from the apple data. 
    
    if English_char(name): # The code in the next couple of lines is different from above, but does the same thing.
        google_data_eng.append(row)
    else:
        google_data_non_eng.append(row)

print("the number of Google apps in English:", len(google_data_eng))
print("the number of Google apps in non-English:", len(google_data_non_eng))

the number of Google apps in English: 9614
the number of Google apps in non-English: 45


## 5. Clean data - free vs. non-free apps (remove non-free apps)
So far in the data cleaning process, I've:
- Removed inaccurate data
- Removed duplicate app entries
- Removed non-English apps

In this particular project, I will need to isolate only the free apps for the analysis.

In [29]:
### Google data ###
print(google_data_header)
print('\n')
explore_data(google_data_eng, 0, 3, True) # which column indicates free vs. non-free in google data?
## index number 7 (8th column) is 'Price'

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


In [30]:
google_data_eng_free = []

for row in google_data_eng:
    price = row[7]
    if price == '0':
        google_data_eng_free.append(row)
        
print("the number of free English Google apps:", len(google_data_eng_free))

the number of free English Google apps: 8864


In [31]:
### Apple data ###
print(apple_data_header)
print('\n')
explore_data(apple_data_eng, 0, 3, True) # which column indicates free vs. non-free in apple data?
## index number 4 (5th column) is 'Price'

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 6183
Number of columns: 16


In [32]:
apple_data_eng_free = []

for row in apple_data_eng:
    price = row[4]
    if price == '0.0':
        apple_data_eng_free.append(row)
        
print("the number of free English Apple apps:", len(apple_data_eng_free))

the number of free English Apple apps: 3222


There are 8864 Android apps and 3222 iOS apps, which should be enough for the analysis.

## 6. Data Analysis I - Identify most common Apps by genre.
So far, I've cleaned the data by:
- Removing inaccurate data
- Removing duplicate app entries
- Removing non-English apps
- Isolating the free apps

The goal is to determine the kinds of apps that are likely to attract more users (as the revenue is highly influenced by the number of people using apps).

Begin the analysis by getting a sense of what are the most common genres for each market. For this, I'll need to build frequency tables for a few columns in the data sets.

First, inspect both data sets and identify the columns I could use to generate frequency tables to find out what are the most common genres in each market.

In [33]:
## Apple / iOS
print(apple_data_header)
print('\n')
explore_data(apple_data_eng_free, 0, 3, True) 
## 'prime_genre': index number -5 (5th from the end)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 3222
Number of columns: 16


In [34]:
## Google / android
print(google_data_header)
print('\n')
explore_data(google_data_eng_free, 0, 3, True) 
## 'Genres': index number -4 (4th from the end)
## 'Category': index number 1 (2nd from the beginning)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 8864
Number of columns: 13


Build two functions I can use to analyze the frequency tables:

- One function to generate frequency tables that show percentages
- Another function to display the percentages in a descending order

In [36]:
# This freq_table function takes two inputs: dataset (which is expected to be a list of lists) and index (which is expected to be an integer).
# This returns the frequency table (as a dictionary) for any column I want. The frequencies should also be expressed as percentages.
    
def freq_table(dataset, index):
    table = {}
    total_apps = 0 # We need this to calculate percentages (below)
    for row in dataset:
        total_apps += 1
        value = row[index]
        if value in table:
            table[value] +=1
        else:
            table[value] = 1
            
    table_percentage = {}
    for key in table: 
        percentage = (table[key] / total_apps)*100
        table_percentage[key] = percentage
        
    return table_percentage

# This display_table function uses the freq_table function above. 
# Transforms the frequency table into a list of tuples, then sorts the list in a descending order.

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key) # converts the dictionary into tuple because sorted() doesn't work for dictionary.
        table_display.append(key_val_as_tuple) 

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

### Apple/iOS apps: "prime_genre" column
The most frequent genre among the free English apps is Games (58% of the free English apps), followed by Entertainment (7.9%). Photo & Video (5%), Education (3.7%), Social Networking (3.3%), Shopping (2.6%), Utilities (2.5%), and so on. 

One observation I could make is that free English iOS apps are mostly for fun (games, entertainment, and photo & video are the top 3 most frequent genres), while apps that are for practical purposes (e.g., shopping, utilities) are more rare. But this is just a frequency of apps' genres; this does not reflect the number of users for these apps. In other words, demand for these genres may be different. 

In [38]:
display_table(apple_data_eng_free, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


### Google/android apps: "Category" column
The most frequent category among the free English apps is Family (18.9% of the free English apps), followed by Game (9.7%), Tools (8.5%), Business (4.6%), Lifestyle (3.9%), Productivity (3.9%), Finance (3.7%), Medical (3.5%) and so on. 

There are some entertainment apps (e.g., Family - though this category doesn't seem to exist in the [store](https://play.google.com/store/apps?hl=en_US&gl=US) now, Game). But there are also practicel apps (Tools, Business, Lifestyle, Productivity, Finance, Medical) 

In [39]:
display_table(google_data_eng_free, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

### Google/android apps: "Genres" column
The distribution of the app 'genre' frequency is much more granular (i.e., there is no app genre that is much more frequent than others) compared to the 'Category' column. 

In [40]:
display_table(google_data_eng_free, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

### Overall

Based on the freuquency information of the app genres/categories, I could make an observation that iOS apps are dominated by 'for fun' apps (e.g., games, entertainment) while android apps are both 'for fun' and 'for practical' purposes (e.g., tools, business, education). 

Now, I'd like to get an idea about the kind of apps with the most users.



## 7. Data Analysis II - Identify the kind of apps (genres) with the most users.
In order to find out what genres have the most users, I will calculate the average number of installs for each app genre. For the Google Play data set, I can find the number of installs in the `Installs` column, but this information is missing for the App Store data set. So, for App Store data set, I'll take the total number of user ratings as a proxy, which can be found in the `rating_count_tot` app.

### The most popular app genre for iOS apps (English free apps)
Start with calculating the average number of user ratings per app genre on the App Store. 
To do that, I'll need to:

- Isolate the apps of each genre.
- Sum up the user ratings for the apps of that genre.
- Divide the sum by the number of apps belonging to that genre (not by the total number of apps).

In [41]:
# First, generate a frequency table for the prime_genre column to get the unique app genres.
# using the freq_table() function from above. 

iOS_genre = freq_table(apple_data_eng_free, -5)
print(iOS_genre)

{'Social Networking': 3.2898820608317814, 'Photo & Video': 4.9658597144630665, 'Games': 58.16263190564867, 'Music': 2.0484171322160147, 'Reference': 0.5586592178770949, 'Health & Fitness': 2.0173805090006205, 'Weather': 0.8690254500310366, 'Utilities': 2.5139664804469275, 'Travel': 1.2414649286157666, 'Shopping': 2.60707635009311, 'News': 1.3345747982619491, 'Navigation': 0.186219739292365, 'Lifestyle': 1.5828677839851024, 'Entertainment': 7.883302296710118, 'Food & Drink': 0.8069522036002483, 'Sports': 2.1415270018621975, 'Book': 0.4345127250155183, 'Finance': 1.1173184357541899, 'Education': 3.662321539416512, 'Productivity': 1.7380509000620732, 'Business': 0.5276225946617008, 'Catalogs': 0.12414649286157665, 'Medical': 0.186219739292365}


In [42]:
## get the average number of ratings for each app genre. 
## numerator = total number of user ratings for a particular genre (e.g., Social Networking)
## demonimator = total number of unique genre

for genre in iOS_genre:
    total = 0 # This stores the sum of user ratings (the number of ratings, not the actual ratings) specific to each genre.
    len_genre = 0 # This stores the number of apps specific to each genre.
    
    for app in apple_data_eng_free:
        genre_app = app[-5]
        
        if genre_app == genre:
            n_ratings = float(app[5]) #save the number of user ratings ('rating_count_tot') as a float
            total += (n_ratings) # Add the number of user ratings to the total variable.
            len_genre += 1 # Increment the len_genre variable by 1
            
    mean_n_ratings = total / len_genre
    print(genre, ':', mean_n_ratings)  

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


Judged by the average number of user ratings of apps in each genre, the iOS English free app genre with the most users (the most popular) is 
- Navigation (average of 86090 user ratings per app), 
- Reference (average of 74942 user ratings per app), 
- Social Networking (average of 71548 user ratings per app). 
- Music (average of 57326 user ratings per app).
- Weather (average of 52279 user ratings per app).

Let's explore the distribution of the number of user ratings for the *Navigation* category. 

In [43]:
for app in apple_data_eng_free:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5])

## As we can see here, the Navigation category is dominated by 'Waze' and 'Google Maps'

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


Let's explore the distribution of the number of user ratings for the *Reference* category. 

In [44]:
for app in apple_data_eng_free:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

## As we can see here, the Navigation category is dominated by 'Bible' and dictionary/translation apps. 

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


Let's explore the distribution of the number of user ratings for the *Social Networking* category.

In [45]:
for app in apple_data_eng_free:
    if app[-5] == 'Social Networking':
        print(app[1], ':', app[5])

## As we can see here, the Navigation category is dominated by 'Facebook'.

Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie Talkie : 27662
Miitomo : 2

As demonstrated above, be careful when interpreting the average number of ratings for a particular category, as a few popular apps (e.g., Facebook in Social Networking) could impact the average, skewing the results. 

As seen in the results of the Reference category, the translation/dictionary apps are very popular. This could suggest that we may be able to integrate this type of function in a popular book app, making it possible to achieve differnet functions within the app. This might lead to more profits. 

### The most popular app genre for Android apps (English free apps)

By looking at the 'Installs' column, I could see how many installs there are for each app. However, the problem here is that the install numbers don't seem precise enough — I can see that most values are open-ended (100+, 1,000+, 5,000+, etc.) See below. 

In [46]:
# the frequency (percentage) of the Installs column (showing the number of times the app has been installed)
display_table(google_data_eng_free,5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


Because I'm interested in which app genres attract the most userts, I don't need very precise number of installs. 
So I'm going to leave the numbers as they are, which means that I'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on. 

To perform computations, however, I'll need to convert each install number from string to float. This means I need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. 

In [47]:
# Generate a frequency table for the Category columnn of the Google Play data set to get the unique app genres.
categories_google = freq_table(google_data_eng_free, 1)


for category in categories_google:
    total = 0 # This variable stores the sum of installs specific to each genre.
    len_category = 0 # This variable stores the number of apps specific to each genre. 
    
    for app in google_data_eng_free:
        category_app = app[1]
        
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace('+', '')
            n_installs = n_installs.replace(',', '')
            n_installs = float(n_installs)
            
            total += n_installs
            len_category += 1
    
    mean_n_installs = total / len_category
    print(category, ':', mean_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

Popular google apps are:
- COMMUNICATION genre: 38456119 installs/app
- VIDEO_PLAYERS: 24727872 installs/app
- SOCIAL genre: 23253652 installs/app
- GAME genre: 15588015 installs/app
- BOOKS_AND_REFERENCE: 8767811

Explore one type genre further: BOOKS_AND_REFERENCE

In [48]:
## very popular apps
for app in google_data_eng_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


In [49]:
## mid-popular apps
for app in google_data_eng_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

# Conclusions

In this project, I analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

Social networking/communication apps are in general popular and attact many users, though a few apps are much more popular than other apps in this type of genre. Other 'for fun' type of apps (e.g., game, music, etc.) and 'pratical' apps (e.g., dictionary, translation) seem to attract many uses for both iOS and android apps as well. 